chirp/drivers/mml_jc8810.py

```diff
@@ -0,0 +1,1070 @@
+DTMFST_LIST = ["Off", "KeyBoard Side Tone", "ANI Side Tone", "KB ST + ANI ST"]
+DUALTX_LIST = ["Off", "A", "B"]
+ENCRYPT_LIST = ["Off", "DCP1", "DCP2", "DCP3"]
+LANGUAGE_LIST = ["English", "Chinese"]
+MDF_LIST = ["Channel + Name", "Channel + Frequency"]
+MENUQUIT_LIST = ["%s seconds" % x for x in range(5, 55, 5)] + ["60 seconds"]
+OFF1TO9_LIST = ["Off"] + ["%s" % x for x in range(1, 10)]
+PONMSG_LIST = ["Logo", "Voltage"]
+PTTID_LIST = ["Off", "BOT", "EOT", "Both"]
+PTTIDCODE_LIST = ["%s" % x for x in range(1, 16)]
+PTTLT_LIST = ["None", "100 seconds"] + \
+             ["%s seconds" % x for x in range(200, 1200, 200)]
+QTSAVE_LIST = ["All", "TX", "RX"]
+RPSTE_LIST = ["Off"] + ["%s ms" % x for x in range(100, 1100, 100)]
+SAVE_LIST = ["Off", "Normal", "Super", "Deep"]
+SCREV_LIST = ["Time (TO)", "Carrier (CO)", "Search (SE)"]
+TAILCODE_LIST = ["55 Hz", "62.5 Hz"]
+TONERXEND_LIST = ["Off", "MDC-1200"]
+TONE_LIST = ["1000 Hz", "1450 Hz", "1750 Hz", "2100 Hz"]
+TOT_LIST = ["Off"] + ["%s seconds" % x for x in range(30, 270, 30)]
+VOX_LIST = ["Off"] + ["%s" % x for x in range(1, 10)]
+VOXD_LIST = ["%s seconds" % str(x / 10) for x in range(5, 21)]
+WORKMODE_LIST = ["VFO Mode", "Channel Mode"]
+
+SKEY_CHOICES = ["FM Radio", "Flashlight", "TX Power Level",
+                "NOAA Weather", "Scan", "Search"]
+SKEY_VALUES = [0x07, 0x08, 0x0A, 0x0C, 0x1C, 0x1D]
+
+SKEY2SP_CHOICES = ["PTT B"] + SKEY_CHOICES
+SKEY2SP_VALUES = [0x01] + SKEY_VALUES
+
+SETTING_LISTS = {
+    "abr": ABR_LIST,
+    "almod": ALMODE_LIST,
+    "autolk": AUTOLK_LIST,
+    "dtmfoff": DTMFSPEED_LIST,
+    "dtmfon": DTMFSPEED_LIST,
+    "dtmfst": DTMFST_LIST,
+    "dualtx": DUALTX_LIST,
+    "encrypt": ENCRYPT_LIST,
+    "language": LANGUAGE_LIST,
+    "mdfa": MDF_LIST,
+    "mdfb": MDF_LIST,
+    "menuquit": MENUQUIT_LIST,
+    "ponmsg": PONMSG_LIST,
+    "pttid": PTTID_LIST,
+    "pttlt": PTTLT_LIST,
+    "qtsave": QTSAVE_LIST,
+    "rpste": RPSTE_LIST,
+    "rptrl": RPSTE_LIST,
+    "rxendtail": TONERXEND_LIST,
+    "save": SAVE_LIST,
+    "scode": PTTIDCODE_LIST,
+    "screv": SCREV_LIST,
+    "tailcode": TAILCODE_LIST,
+    "tone": TONE_LIST,
+    "tot": TOT_LIST,
+    "vox": OFF1TO9_LIST,
+    "voxd": VOXD_LIST,
+    "workmode": WORKMODE_LIST
+    }
+
+
+def _enter_programming_mode(radio):
+    serial = radio.pipe
+
+    exito = False
+    for i in range(0, 5):
+        serial.write(radio._magic)
+        ack = serial.read(1)
+
+        try:
+            if ack == CMD_ACK:
+                exito = True
+                break
+        except errors.RadioError:
+            LOG.debug("Attempt #%s, failed, trying again" % i)
+            pass
+
+    # check if we had EXITO
+    if exito is False:
+        msg = "The radio did not accept program mode after five tries.\n"
+        msg += "Check you interface cable and power cycle your radio."
+        raise errors.RadioError(msg)
+
+    try:
+        serial.write(b"F")
+        ident = serial.read(8)
+    except errors.RadioError:
```

Comment:
These serial calls can't ever raise the chirp-specific `errors.RadioError` so this will never happen. This should probably be `except Exception` to cache OS errors reading or writing to the pipe.